arrabiata/src/interpreter.rs

```diff
@@ -0,0 +1,962 @@
+//! correctly computed.
+//! In the implementation, the accumulator will be computed after the challenges
+//! and the commitments. Note that the accumulator must also be aggregated, and
+//! the aggregation must be performed by the verifier at the next step.
+//!
+//! The methods `save` and `load` will accept as arguments only a column that is
+//! included in the permutation argument. For instance, `save_poseidon_state`
+//! will only accept columns with index 3, 4 and 5, where the
+//! `load_poseidon_state` will only accepts columns with index 0, 1 and 2.
+//!
+//! The permutations values will be saved in public values, and will contain the
+//! index of the row. The permutation values will be encoded with a 32 bits
+//! value (u32) as we can suppose a negligible probability that a user will use
+//! more than 2^32 rows.
+//!
+//! The permutation argument also generates constraints that will be
+//! homogeneized with the gadget constraints.
+//!
+//! Note all rows might require to use the permutation argument. Therefore, a
+//! selector will be added to activate/deactivate the permutation argument.
+//! When a method calls `save` or `load`, the selector will be activated. By
+//! default, the selector will be deactivated.
+//!
+//! TBD:
+//! - number of columns
+//! - accumulator column
+//! - folding of the permutation argument
+//!
+//! TBD/FIXME: do we use a additive permutation argument to increase the number
+//! of columns we can perform the permutation on?
+//!
+//! TBD/FIXME: We can have more than one permutation argument. For instance, we
+//! can have a permutation argument for the columns 0, 1, 2, 3 and one for the
+//! columns 4, 5, 6, 7. It can help to decrease the degree.
+//!
+//! ## Fiat-Shamir challenges
+//!
+//! The challenges sent by the verifier must also be simulated by the IVC
+//! circuit.
+//!
+//! For a step `i + 1`, the challenges of the step `i` must be computed by the
+//! verifier, and check that it corresponds to the ones received as a public
+//! input.
+//!
+//! TBD/FIXME: specify. Might require foreign field arithmetic.
+//!
+//! TBD/FIXME: do we need to aggregate them for the end?
+//!
+//! ## Folding
+//!
+//! Constraints must be homogeneized for the folding scheme.
+//! Homogeneising a constraint means that we add a new variable (called "U" in
+//! Nova for instance) that will be used to homogeneize the degree of the monomials
+//! forming the constraint.
+//! Next to this, additional information, like the cross-terms and the error
+//! terms must be computed.
+//!
+//! This computation depends on the constraints, and in particular on the
+//! monomials describing the constraints.
+//! The computation of the cross-terms and the error terms happen after the
+//! witness has been built and the different arguments like the permutation or
+//! lookup have been done. Therefore, the interpreter must provide a method to
+//! compute it, and the constraints should be passed as an argument.
```

Comment:
I would give the symbolic cross term (computed at setup time) as an argument, rather than the contraints